project14-Whack-a-Penguin/GameScene.swift

```diff
@@ -9,12 +9,118 @@ import SpriteKit
+            self.creatEnemy()
+        }
         
     }
     
     override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
+        guard let touch = touches.first else { return }
+        let location = touch.location(in: self)
+        let tappedNodes = nodes(at: location)
+        
+        for node in tappedNodes {
+            guard let whackSlot = node.parent?.parent as? WhackSlot else {
+                continue }
+            if !whackSlot.isVisible { continue }
+            if whackSlot.isHit { continue }
+            whackSlot.hit()
+   
+            
+            if node.name == "charFriend" {
+                // they shouldn't have whacked this penguin
+                 score -= 5
+                run(SKAction.playSoundFileNamed("whackBad.caf", waitForCompletion: false))
+    
+            } else if node.name == "charEnemy" {
+                //they should have whacked this one                
+                whackSlot.charNode.xScale = 0.85
+                whackSlot.charNode.yScale = 0.85
 
+                score += 1
+                
+                run(SKAction.playSoundFileNamed("whack.caf", waitForCompletion: false))
+                
+            }
+        }
+    }
+    
+    private func createSlot(at position: CGPoint) {
+        let slot = WhackSlot()
+        slot.configure(at: position)
+        addChild(slot)
+        slots.append(slot)
+    }
+    private func creatEnemy() {
+        numRounds += 1
+        
+        if numRounds >= 30 {
+            for slot in slots {
+                slot.hidePenguin()
+            }
+            let gameOver = SKSpriteNode (imageNamed: "gameOver")
+            gameOver.position = CGPoint(x: 512, y: 384)
+            gameOver.zPosition = 1
+            addChild(gameOver)
+            let totalScore = SKLabelNode (fontNamed: "Chalkduster")
+            totalScore.text = gameScore.text
+            totalScore.position = CGPoint(x: 380, y: 300)
+            totalScore.zPosition = 1
+            totalScore.horizontalAlignmentMode = .left
+            totalScore.fontSize = 50
+            addChild(totalScore)
+            run(SKAction.playSoundFileNamed("gameOver.m4a", waitForCompletion: false))
+            return
+        }
+        popupTime *= 0.991
+        slots.shuffle()
+        slots[0].show(hideTime: popupTime)
+        
+        if Int.random(in: 0...12) > 4 { slots[1].show(hideTime: popupTime)}
+        if Int.random(in: 0...12) > 8 { slots[2].show(hideTime: popupTime)}
+        if Int.random(in: 0...12) > 10 { slots[3].show(hideTime: popupTime)}
+        if Int.random(in: 0...12) > 11 { slots[4].show(hideTime: popupTime)}
+        
+        let minDelay = popupTime / 2.0
+        let maxDelay = popupTime * 2
+        let delay = Double.random(in: minDelay...maxDelay)
+        
+        DispatchQueue.main.asyncAfter(deadline: .now() + delay) {
+            [weak self] in
+            guard let self = self else { return }
+            self.creatEnemy()
```

Comment:
In this case you don't need guard against self since you have just one task. It could just read `self?.createEnemy()`.

Reply:
I see. 
Ok Thank you!